### DISCUSSION TITLE
Added callback support to MSC SCSI command execution

### DISCUSSION THREAD

**Discussion by Hustin7** [hathach/tinyusb#1743] [Show and tell]
I noticed that many calls to tuh_msc_scsi_command() supply a callback but there seems to be no provision to actually use the callback. Not sure if this is a hack or not (?) but in my code I added support for callbacks. Feel free to criticize or use:

```
void msc_complete_cb(tuh_xfer_t* xfer)
{
  msch_interface_t* p_msc = (msch_interface_t*) xfer->user_data;
  p_msc->complete_cb(xfer->daddr, &(p_msc->cbw), &(p_msc->csw));
}

bool tuh_msc_scsi_command(uint8_t dev_addr, msc_cbw_t const* cbw, void* data, tuh_msc_complete_cb_t complete_cb)
{
  msch_interface_t* p_msc = get_itf(dev_addr);
  static tuh_xfer_t xfer;

  TU_VERIFY(p_msc->configured);

  // TODO claim endpoint

  p_msc->cbw = *cbw;
  p_msc->stage = MSC_STAGE_CMD;
  p_msc->buffer = data;
  p_msc->complete_cb = complete_cb;

  //!! Added support for callbacks
  if (complete_cb)
  {
    xfer.daddr = dev_addr;
    xfer.ep_addr = p_msc->ep_out;
    xfer.buffer = (uint8_t*) &p_msc->cbw;
    xfer.buflen = sizeof(msc_cbw_t);
    xfer.complete_cb = msc_complete_cb;
    xfer.user_data = (uintptr_t) p_msc;
    TU_ASSERT(usbh_edpt_xfer_with_callback(xfer.daddr, xfer.ep_addr, xfer.buffer, (uint16_t) xfer.buflen, xfer.complete_cb, xfer.user_data));
  }
  else
  {
    TU_ASSERT(usbh_edpt_xfer(dev_addr, p_msc->ep_out, (uint8_t*) &p_msc->cbw, sizeof(msc_cbw_t)));
  }

  return true;
}
```

**Comment by hathach** [comment]
can you elaborate in more details. Which **specific** function supply callback and not use it ? and how your snippet help with that.

**Comment by Hustin7** [comment]
There are three functions that are called with callbacks and subsequently call tuh_msc_scsi_command:

tuh_msc_read_capacity (called by config_test_unit_ready_complete)
tuh_msc_test_unit_ready (called by config_get_maxlun_complete and config_request_sense_complete)
tuh_msc_request_sense (called by config_test_unit_ready_complete)

In addition, there are others that take callbacks that call tuh_msc_scsi_command:

tuh_msc_inquiry
tuh_msc_read10
tuh_msc_write10

In all of these cases, if tuh_msc_scsi_command doesn't support handling the callbacks then the callback would never get called. It only gets thrown away.

**Reply by hathach** [reply]
tuh_msc_scsi_command() already support complete callback using its own driver complete_cb. 

```
p_msc->complete_cb = complete_cb;
```

usbh_edpt_xfer_with_callback() is not sufficient for use with msc, because 1 msc command involved both IN + OUT endpoint with 3 phase (command, data, status).

**Reply by Hustin7** [reply]
tuh_msc_scsi_command() sets p_msc->complete_cb, but nothing subsequently calls it. The call to usbh_edpt_xfer() throws it away. Where does p_msc->complete_cb actually get used?

**Reply by hathach** [reply]
have you tried running tuh_msc_scsi_command() with complete callback and the callback isn't fired ? If that is the case, please post your example along with debug LOG=2 here for analysis.

**Comment by Hustin7** [comment]
When I turn on debug logging I can't get anything to mount at all, so I used some more lightweight logging.

Now I understand what's going on. The code is using get_itf() to make the callback persistent across routine calls, such that  a call to usbh_edpt_xfer_with_callback isn't necessary.

**Reply by hathach** [reply]
thanks for confirmation, glad that all works as it should. The codebase grow a bit too fast, it is always good to run it on actual hardware first for testing/observing.
